stream.go

```diff
@@ -0,0 +1,451 @@
+		sr.fixedArgs = []string{"GROUP", sr.opts.Group, sr.opts.Consumer}
+	} else {
+		sr.cmd = "XREAD"
+		sr.fixedArgs = nil
+	}
+
+	if sr.opts.Count > 0 {
+		sr.fixedArgs = append(sr.fixedArgs, "COUNT", strconv.Itoa(sr.opts.Count))
+	}
+
+	if sr.opts.Block >= 0 {
+		sec := int(sr.opts.Block / time.Millisecond)
+		sr.fixedArgs = append(sr.fixedArgs, "BLOCK", strconv.Itoa(sec))
+	}
+
+	if sr.opts.Group != "" && sr.opts.Noack {
+		sr.fixedArgs = append(sr.fixedArgs, "NOACK")
+	}
+
+	sr.streams = make([]string, 0, len(sr.opts.Streams))
+	sr.ids = make(map[string]string, len(sr.opts.Streams))
+	for stream, id := range sr.opts.Streams {
+		sr.streams = append(sr.streams, stream)
+
+		if id != nil {
+			sr.ids[stream] = id.String()
+		} else if sr.cmd == "XREAD" {
+			sr.ids[stream] = "$"
+		} else if sr.cmd == "XREADGROUP" {
+			sr.ids[stream] = ">"
+		}
+	}
+
+	// set to nil so we don't accidentally use it later, since the user could have changed
+	// the map after using the reader.
+	sr.opts.Streams = nil
+
+	sr.fixedArgs = append(sr.fixedArgs, "STREAMS")
+	sr.fixedArgs = append(sr.fixedArgs, sr.streams...)
+
+	// preallocate space for all arguments passed to Cmd
+	sr.args = make([]string, 0, len(sr.fixedArgs)+len(sr.streams))
+}
+
+func (sr *streamReader) do() error {
+	sr.args = append(sr.args[:0], sr.fixedArgs...)
+
+	for _, s := range sr.streams {
+		sr.args = append(sr.args, sr.ids[s])
+	}
+
+	return sr.c.Do(Cmd(&sr.unmarshaler, sr.cmd, sr.args...))
+}
+
+// Err implements the StreamReader interface.
+func (sr *streamReader) Err() error {
+	return sr.err
+}
+
+// Next implements the StreamReader interface.
+func (sr *streamReader) Next() (stream string, entries []StreamEntry, ok bool) {
+	if sr.err != nil {
+		return "", nil, false
+	}
+
+	if len(sr.unmarshaler.A) == 0 {
+		if sr.err = sr.do(); sr.err != nil {
+			return "", nil, false
+		}
+
+		if len(sr.unmarshaler.A) == 0 {
+			return "", nil, false
+		}
+	}
+
+	sre := sr.unmarshaler.A[len(sr.unmarshaler.A)-1]
+	sr.unmarshaler.A = sr.unmarshaler.A[:len(sr.unmarshaler.A)-1]
+
+	// do not update the ID for XREADGROUP when we are not reading unacknowledged entries.
+	if sr.cmd == "XREAD" || (sr.cmd == "XREADGROUP" && sr.ids[sre.stream] != ">") {
+		sr.ids[sre.stream] = sre.entries[len(sre.entries)-1].ID.String()
+	}
+
+	return sre.stream, sre.entries, true
+}
+
+type streamReaderEntry struct {
+	stream  string
+	entries []StreamEntry
+}
+
+// streamReaderUnmarshaler implements unmarshaling of XREAD and XREADGROUP responses.
+//
+// The logic is not part of streamReader so that users can not use streamReader directly
+// as a resp.Unmarshaler.
+type streamReaderUnmarshaler struct {
+	A []streamReaderEntry
+
+	streams map[string]string // map of stream names to avoid string allocations
+}
+
+var _ resp.Unmarshaler = (*streamReaderUnmarshaler)(nil)
+
+// UnmarshalRESP implements the resp.Unmarshaler interface.
+func (s *streamReaderUnmarshaler) UnmarshalRESP(br *bufio.Reader) error {
```

Comment:
same question as with `StreamEntry.UnmarshalRESP`: do we know how much time is actually being saved with all of this code, versus just using `Any`?